controller/volume_attachment_controller.go

```diff
@@ -0,0 +1,529 @@
+	// Volume controller start detach volume -> state become detached
+	// VA controller set Vol.Spec.NodeID to node-2
+	// If vol.Status.State is detached, VA controller VA.Attachment.Status.NodeID = ""; VA.Attachment.Status.Parameter = {}; VA.Attachment.Status.Type = ""
+	// Volume controller start attach -> state becomes attached to node-2
+	// If vol.Status.State is attached, VA controller VA.Attachment.Status.NodeID = node-2; VA.Attachment.Status.Parameter = {xx}; VA.Attachment.Status.Type = "xx"
+	//}
+
+	// TODO: consider handle the csi-attacher ticket for RWX volume differently
+
+	if vol.Status.CurrentNodeID == "" || vol.Status.State != longhorn.VolumeStateAttached {
+		attachmentStatus.Satisfied = false
+		attachmentStatus.Conditions = types.SetCondition(
+			attachmentStatus.Conditions,
+			longhorn.AttachmentStatusConditionTypeSatisfied,
+			longhorn.ConditionStatusFalse,
+			"", // reason should be a code indicating the error type
+			"",
+		)
+
+		// TODO: check if the engine image is ready on the node
+		// check if the node is down
+		// to set the condition for the client to consume
+		return nil
+	}
+
+	if attachmentSpec.NodeID != vol.Status.CurrentNodeID {
+		attachmentStatus.Satisfied = false
+		attachmentStatus.Conditions = types.SetCondition(
+			attachmentStatus.Conditions,
+			longhorn.AttachmentStatusConditionTypeSatisfied,
+			longhorn.ConditionStatusFalse,
+			"",
+			fmt.Sprintf("cannot satisify this attachment request because the volume already attached to different node %v ", vol.Status.CurrentNodeID),
+		)
+		return nil
+	}
+
+	if vol.Status.CurrentNodeID == attachmentSpec.NodeID && vol.Status.State == longhorn.VolumeStateAttached {
+		if !verifyAttachmentParameters(attachmentSpec.Parameters, vol) {
+			attachmentStatus.Satisfied = false
+			attachmentStatus.Conditions = types.SetCondition(
+				attachmentStatus.Conditions,
+				longhorn.AttachmentStatusConditionTypeSatisfied,
+				longhorn.ConditionStatusFalse,
+				"",
+				fmt.Sprintf("volume %v has already attached to node %v with incompatible parameters", vol.Name, vol.Status.CurrentNodeID),
+			)
+			return nil
+		}
+		attachmentStatus.Satisfied = true
+		attachmentStatus.Conditions = types.SetCondition(
+			attachmentStatus.Conditions,
+			longhorn.AttachmentStatusConditionTypeSatisfied,
+			longhorn.ConditionStatusTrue,
+			"",
+			"",
+		)
+	}
+	return nil
+}
+
+func verifyAttachmentParameters(parameters map[string]string, vol *longhorn.Volume) bool {
+	disableFrontendString, ok := parameters["disableFrontend"]
+	if !ok || disableFrontendString == longhorn.FalseValue {
+		return vol.Spec.DisableFrontend == false
+	} else if disableFrontendString == longhorn.TrueValue {
+		return vol.Spec.DisableFrontend == true
+	}
+	return true
+}
+
+func setAttachmentParameter(parameters map[string]string, vol *longhorn.Volume) {
+	disableFrontendString, ok := parameters["disableFrontend"]
+	if !ok || disableFrontendString == longhorn.FalseValue {
```

Comment:
Why is `vol.Spec.DisableFrontend` set to `false` rather than `true` when `parameters=map[string]string{}` (I mean reset operation when there is no attachment ticket)?

Reply:
I am assuming the default behavior is to enable frontend unless the client asks for disabling it. This is previous behavior when attach calls without DisableFrontend